src/algebra/quaternion_basis.lean

```diff
@@ -0,0 +1,166 @@
+  congr',
+  rw [←q₁_i_mul_j, ←q₂_i_mul_j],
+  congr'
+end
+
+variables (R)
+
+/-- There is a natural quaternionic basis for the `quaternion_algebra`. -/
+@[simps i j k]
+protected def self : basis ℍ[R,c₁,c₂] c₁ c₂ :=
+{ i := ⟨0, 1, 0, 0⟩,
+  i_mul_i := by { ext; simp },
+  j := ⟨0, 0, 1, 0⟩,
+  j_mul_j := by { ext; simp },
+  k := ⟨0, 0, 0, 1⟩,
+  i_mul_j := by { ext; simp },
+  j_mul_i := by { ext; simp } }
+
+variables {R}
+
+instance : inhabited (basis ℍ[R,c₁,c₂] c₁ c₂) := ⟨basis.self R⟩
+
+variables (q : basis A c₁ c₂)
+include q
+
+attribute [simp] i_mul_i j_mul_j i_mul_j j_mul_i
+
+@[simp] lemma i_mul_k : q.i * q.k = c₁ • q.j :=
+by rw [←i_mul_j, ←mul_assoc, i_mul_i, smul_mul_assoc, one_mul]
+
+@[simp] lemma k_mul_i : q.k * q.i = -c₁ • q.j :=
+by rw [←i_mul_j, mul_assoc, j_mul_i, mul_neg_eq_neg_mul_symm, i_mul_k, neg_smul]
+
+@[simp] lemma k_mul_j : q.k * q.j = c₂ • q.i :=
+by rw [←i_mul_j, mul_assoc, j_mul_j, mul_smul_comm, mul_one]
+
+@[simp] lemma j_mul_k : q.j * q.k = -c₂ • q.i :=
+by rw [←i_mul_j, ←mul_assoc, j_mul_i, neg_mul_eq_neg_mul_symm, k_mul_j, neg_smul]
+
+@[simp] lemma k_mul_k : q.k * q.k = -((c₁ * c₂) • 1) :=
+by rw [←i_mul_j, mul_assoc, ←mul_assoc q.j _ _, j_mul_i, ←i_mul_j,
+  ←mul_assoc, mul_neg_eq_neg_mul_symm, ←mul_assoc, i_mul_i, smul_mul_assoc, one_mul, neg_mul_eq_neg_mul_symm,
+  smul_mul_assoc, j_mul_j, smul_smul]
+
+/-- Intermediate result used to define `quaternion_algebra.basis.lift_hom`. -/
+def lift (x : ℍ[R,c₁,c₂]) : A :=
+algebra_map R _ x.re + x.im_i • q.i + x.im_j • q.j + x.im_k • q.k
+
+lemma lift_zero : q.lift (0 : ℍ[R,c₁,c₂]) = 0 := by simp [lift]
+lemma lift_one : q.lift (1 : ℍ[R,c₁,c₂]) = 1 := by simp [lift]
+lemma lift_add (x y : ℍ[R,c₁,c₂]) : q.lift (x + y) = q.lift x + q.lift y :=
+by { simp [lift, add_smul], abel }
+lemma lift_mul (x y : ℍ[R,c₁,c₂]) : q.lift (x * y) = q.lift x * q.lift y :=
+begin
+  simp only [lift, algebra.algebra_map_eq_smul_one],
+  simp only [add_mul],
+  simp only [add_mul, mul_add, smul_mul_assoc, mul_smul_comm, one_mul, mul_one,
+    ←algebra.smul_def, smul_add, smul_smul],
+  simp only [i_mul_i, j_mul_j, i_mul_j, j_mul_i, i_mul_k, k_mul_i, k_mul_j, j_mul_k, k_mul_k],
+  simp only [smul_smul, smul_neg, sub_eq_add_neg, add_smul, ←add_assoc, mul_neg_eq_neg_mul_symm, neg_smul],
+  simp only [mul_right_comm _ _ (c₁ * c₂), mul_comm _ (c₁ * c₂)],
+  simp only [mul_comm _ c₁, mul_right_comm _ _ c₁],
+  simp only [mul_comm _ c₂, mul_right_comm _ _ c₂],
+  simp only [←mul_comm c₁ c₂, ←mul_assoc],
+  simp [sub_eq_add_neg, add_smul, ←add_assoc],
+  abel
+end
+
+lemma lift_smul (r : R) (x : ℍ[R,c₁,c₂]) : q.lift (r • x) = r • q.lift x :=
+by simp [lift, mul_smul, ←algebra.smul_def]
+
+/-- A `quaternion_algebra.basis` implies an `alg_hom` from the quaternions. -/
+@[simps]
+def lift_hom : ℍ[R,c₁,c₂] →ₐ[R] A :=
+alg_hom.mk'
+  { to_fun := q.lift,
+    map_zero' := q.lift_zero,
+    map_one' := q.lift_one,
+    map_add' := q.lift_add,
+    map_mul' := q.lift_mul }
+  q.lift_smul
```

Comment:
I was also able to prove this all at once as:
```lean
/-- A `quaternion_algebra.basis` implies an `alg_hom` from the quaternions. -/
def lift_aux : ℍ[R,c₁,c₂] →ₐ[R] A :=
alg_hom.of_linear_map
  (
    (algebra.linear_map R A).comp (re_lm R c₁ c₂) +
    (linear_map.to_span_singleton _ _ q.i).comp (im_i_lm R c₁ c₂) +
    (linear_map.to_span_singleton _ _ q.j).comp (im_j_lm R c₁ c₂) +
    (linear_map.to_span_singleton _ _ q.k).comp (im_k_lm R c₁ c₂))
  (show algebra_map R A 1 + (0 : R) • q.i + (0 : R) • q.j + (0 : R) • q.k = 1,
    by simp only [ring_hom.map_one, zero_smul, add_zero])
  begin
    intros x y,
    dsimp only [linear_map.comp_apply, linear_map.to_span_singleton, linear_map.id_apply,
      re_lm_apply, im_i_lm_apply, im_j_lm_apply, im_k_lm_apply, linear_map.add_apply,
      linear_map.smul_right_apply, algebra.linear_map_apply],
    simp only [algebra.algebra_map_eq_smul_one],
    simp only [add_mul],
    simp only [add_mul, mul_add, smul_mul_assoc, mul_smul_comm, one_mul, mul_one,
      ←algebra.smul_def, smul_add, smul_smul],
    simp only [i_mul_i, j_mul_j, i_mul_j, j_mul_i, i_mul_k, k_mul_i, k_mul_j, j_mul_k, k_mul_k],
    simp only [smul_smul, smul_neg, sub_eq_add_neg, add_smul, ←add_assoc, mul_neg_eq_neg_mul_symm, neg_smul],
    simp only [mul_right_comm _ _ (c₁ * c₂), mul_comm _ (c₁ * c₂)],
    simp only [mul_comm _ c₁, mul_right_comm _ _ c₁],
    simp only [mul_comm _ c₂, mul_right_comm _ _ c₂],
    simp only [←mul_comm c₁ c₂, ←mul_assoc],
    simp [sub_eq_add_neg, add_smul, ←add_assoc],
    abel
  end
```
but it's not really more readable.